saleor/discount/tasks.py

```diff
@@ -1,96 +1,152 @@
 from collections import defaultdict
 from datetime import datetime
-from typing import Dict
+from typing import TYPE_CHECKING, Dict
 
 import graphene
 import pytz
 from celery.utils.log import get_task_logger
-from django.db.models import F, Q
+from django.db.models import Exists, F, OuterRef, Q, QuerySet
 
 from ..celeryconf import app
-from ..graphql.discount.mutations.utils import CATALOGUE_FIELD_TO_TYPE_NAME
+from ..graphql.discount.utils import get_variants_for_predicate
 from ..plugins.manager import get_plugins_manager
-from ..product.tasks import update_products_discounted_prices_of_catalogues_task
-from .models import Sale
-from .utils import CATALOGUE_FIELDS, CatalogueInfo
+from ..product.models import Product, ProductVariant
+from ..product.tasks import update_products_discounted_prices_for_promotion_task
+from .models import Promotion, PromotionRule
+
+if TYPE_CHECKING:
+    from uuid import UUID
 
 task_logger = get_task_logger(__name__)
 
 
 @app.task
-def handle_sale_toggle():
-    """Send the notification about sales toggle and recalculate discounted prcies.
+def handle_promotion_toggle():
+    """Send the notification about promotion toggle and recalculate discounted prices.
 
-    Send the notifications about starting or ending sales and call recalculation
+    Send the notifications about starting or ending promotions and call recalculation
     of product discounted prices.
     """
     manager = get_plugins_manager()
 
-    sales = get_sales_to_notify_about()
-
-    sale_id_to_catalogue_infos, catalogue_infos = fetch_catalogue_infos(sales)
+    staring_promotions = get_starting_promotions()
+    ending_promotions = get_ending_promotions()
 
-    if not sales:
+    if not staring_promotions and not ending_promotions:
         return
 
-    for sale in sales:
-        catalogues = sale_id_to_catalogue_infos.get(sale.id)
-        manager.sale_toggle(sale, catalogues)
+    promotions = staring_promotions | ending_promotions
+    promotion_id_to_variants, product_ids = fetch_promotion_variants_and_product_ids(
+        promotions
+    )
 
-    if catalogue_infos:
+    for staring_promo in staring_promotions:
+        manager.promotion_started(staring_promo)
+
+    for ending_promo in ending_promotions:
+        manager.promotion_ended(ending_promo)
+
+    # DEPRECATED: will be removed in Saleor 4.0.
```

Comment:
Maybe it would be worth adding a gh issue about this, and including to [Saleor4.0 mileston](https://github.com/saleor/saleor/milestone/80)?

Reply:
Added: https://github.com/saleor/saleor/issues/13573#issue-1828825452